src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2749,13 +2813,567 @@ def query(self, filters=None, options=None):
+
+        if not verrors and ZFSKeyFormat(options['key_format']) == ZFSKeyFormat.PASSPHRASE:
+            if any(
+                d['name'] == d['encryption_root']
+                for d in await self.middleware.run_in_thread(
+                    self.query, [
+                        ['id', '^', f'{id}/'], ['encrypted', '=', True],
+                        ['key_format.value', '!=', ZFSKeyFormat.PASSPHRASE.value]
+                    ]
+                )
+            ):
+                verrors.add(
+                    'change_key_options.key_format',
+                    f'{id} has children which are encrypted with HEX/RAW keys. It is not allowed to '
+                    'have HEX/RAW encrypted roots as children for PASSPHRASE encrypted datasets.'
+                )
+            elif id == (await self.middleware.call('systemdataset.config'))['pool']:
+                verrors.add(
+                    'id', f'{id} contains system dataset. Please move system dataset before changing key_format'
+                )
+
+        verrors.check()
+
+        encryption_dict = await self.middleware.call(
+            'pool.dataset.validate_encryption_data', job, verrors, {
+                'enabled': True, 'key_format': options['key_format'], 'passphrase': options.get('passphrase'),
+                'generate_key': options['generate_key'], 'key_file': options.get('key_file'),
+                'pbkdf2iters': options['pbkdf2iters'], 'algorithm': 'on',
+            }, 'change_key_options'
+        )
+
+        verrors.check()
+
+        encryption_dict.pop('encryption')
+        key = encryption_dict.pop('key')
+
+        await self.middleware.call(
+            'zfs.dataset.change_key', id, {
+                'encryption_properties': encryption_dict,
+                'key': key, 'load_key': False,
+            }
+        )
+
+        # TODO: Handle renames of datasets appropriately wrt encryption roots and db - this will be done when
+        #  devd changes are in from the OS end
+        await self.insert_or_update_encrypted_record(
+            {'encryption_key': key, 'key_format': options['key_format'], 'name': id}
+        )
+
+    @accepts(Str('id'))
+    async def inherit_parent_encryption_properties(self, id):
+        """
+        Allows to inherit parent's encryption root discarding it's current encryption settings. It should be noted
+        this can only be done where `id` has an encrypted parent and `id` itself is an encryption root.
+        """
+        ds = await self._get_instance(id)
+        if not ds['encrypted']:
+            raise CallError(f'Dataset {id} is not encrypted')
+        elif not ds['encryption_root'] != id:
+            raise CallError(f'Dataset {id} is not an encryption root')
+        elif ds['locked']:
+            raise CallError('Dataset must be unlocked to perform this operation')
+        elif '/' not in id:
+            raise CallError('This operation is not valid for root datasets')
```

Comment:
Does "not valid" mean "not allowed" or "cannot be done on root datasets"?

Reply:
Yes

Reply:
I would argue that "not valid" sounds like it's an illegal operation so were stopping you while not allowed sounds like some restriction were doing by choice as part of freenas.

Reply:
It’s not possible to do this operation on root datasets. Apologies for the vague comment above

Reply:
But that just says the same thing again.  Being specific avoids the problems with "invalid" or "not allowed":
"Root datasets do not have a parent and cannot inherit encryption settings."